README.md
# proxycheck

## TODO

This is a work in progress -- All options/arguments and functionality are subject to change. That said, everything noted in this README is up-to-date and reflects the current capability (especially the "Usage" and "Examples" sections). 

## Language

Written in C

## Purpose:

A proof-of-concept tool for testing various domains/URLs via a proxy

## Overview

The idea is to provide a way to audit if a domain/URL is accessible, 
given a request being sent. This tool is designed to provide a 
way to (hopefully) quickly get results, knowing there could be a 
huge variety of domains/URLs and spot checking may not be a good 
way to proceed.

This tool is not designed to care about a successful request 
versus a failed request as it relates to an HTTP status result.

At a high level, the tool:

- parses a list of one or more requested domains/URLs
- builds the request into a list of one or more request objects
- establishes a TCP connection to a listening server and port 
(ideally this is a proxy)
- writes the built request through the socket
- reads the response
- processes the response (most notably to derive a status)
- evaluates and outputs the results

The tool evaluates an HTTP status response in a binary fasion 
from the perspective of the request either made it past the 
proxy or it didn't.

There are two high level flows, given a request:
1. CONNECT
2. STANDARD

With a CONNECT flow, the request is conducted by establishing a 
TCP connection with the server (most likely the proxy), sending 
the request through the socket, and retrieving the response. Upon 
obtaining the response, it is evaluated to determine success 
or failure. If the response is anything other than an HTTP 200, 
an error is generated and no further processing on that request 
is attempted. If the CONNECT results in an HTTP 200, then another
request is sent to the upstream server following the STANDARD
flow.

With the STANDARD flow, the request is sent via the previously 
created socket as long as it previously followed a CONNECT flow. 
Otherwise, a new socket is created with the server (most likely 
the proxy). Next, the request is sent through the socket before 
retrieving the response. The response is evaluated against one 
of several different status codes.

It is up to the user to determine what success versus failure is 
from an audit perspective. After all, it can vary on the use case:

Example: You may want to audit whether a list of domains/URLs are 
reachable. If this results in the contrary, it could mean there is 
an issue that needs to be investigated (network, proxy, logical error 
in policy, etc.). In this case, anything other than a success is most 
likely not a good thing.

Example: You may want to audit whether a list of domains/URLs are NOT 
reachable. In this case, if a request is successful, it most likely 
means an issue with the defined proxy policy ACLs. Again, most likely 
not a good thing.

In the end, the user can take the results and determine the best course 
of action, given their use case(s). The results can be fed into a tool 
designed for ingesting events. Further metrics, alerts, etc. can be created. 
Finally, action can be taken from these results if necessary or desired.

## Supported Platforms

Compiles and runs on:

- Unix (such as FreeBSD or variants like OS X)
- Linux (such as CentOS)
- Windows (requires cygwin1.dll for native use via a command prompt) or it can operate via a Cygwin terminal if preferred

## Compile:

```
gcc -o proxycheck proxycheck_main.c proxycheck.c
```
OR
```
gcc -std=<standard> -o proxycheck proxycheck_main.c proxycheck.c

Where <standard> is: c99, c11, etc
```

## Usage

```
Usage:
	proxycheck [options] <arguments>
	proxycheck <arguments> [options]

[Options]

-B [resp-buf]		# The response buffer to allocate for each request. By
			# default, a sufficient buffer is allocated to hold
			# HTTP response status only. If the option is provided
			# and [resp-buf] is omitted, then the buffer will be
			# increased to 512 bytes. Setting the buffer can be
			# useful when using -V to view responses or following
			# redirects (see: -F).
			# [NOTE]: When [resp-buf] is provided, the range is:
			# 10 bytes - 65535 bytes.
			#
-D [interval]		# The amount of time (in seconds) to sleep between
			# processing requests via <proxy-server>. Should
			# [interval] be omitted, the amount of time sleeping
			# is 3600 seconds.
			# [NOTE]: Tool will run as a background daemon process.
			# [NOTE]: If -L is specified, events are written to
			# <request-log>. Otherwise, the events are written to
			# the default log: "proxycheck.log".
			#
-F <resp-status ...>	# Attempt to follow one or more HTTP status codes. Must
			# consist of one or more of the following:
			#	- 301
			#	- 302
			#	- 303
			#	- 307
			# [NOTE]: Valid delimiters are the same as <request>.
			# [NOTE]: Due to where the Location header might be
			# located among the HTTP response headers, using the
			# -B option can be useful to ensure the response buffer
			# is adequate to contain the Location header.
			#
-H <header...>		# One or more headers to include in the request. If
			# using more than one, they should be appended.
			# Supported headers:
			#	- A or a (adds Accept header)
			#	- C or c (adds Connection header)
			#	- H or h (adds Host header)
			#	- U or u (adds User-Agent header)
			# [NOTE]: Some servers expect certain headers (such
			# as a Host header) and might respond with an error
			# (e.g. HTTP 400). In such cases, adding a header is
			# advised to minimize false results.
			#
-I <sec>		# The amount of time (in seconds) to wait for input
			# from a request to be retrieved via <proxy-server>.
			# This timer is reset each time data is read via the
			# socket. Valid range is between: 0 - 120 seconds.
			# When omitted, it defaults to: 10 seconds.
			#
-L <request-log>	# Send stdout and stderr streams to <request-log>. If
			# <request-log> does not exist, it will be created
			# or an error is returned. If <request-log> does
			# exist, it is appended to.
			#
-M <method>		# HTTP method to use for <request>. Must be one of:
			#  - CONNECT
			#  - GET
			#  - HEAD
			#  - OPTIONS
			# If omitted, an HTTP GET will be used. If CONNECT
			# is specified, then HTTP GET will be used for the
			# subsequent request. When CONNECT is specified and
			# a scheme exists in <request>, then HTTP CONNECT
			# will be implied and an HTTP GET will be used for
			# the follow-up request. When <scheme> is HTTPS://
			# and not already present in <request>, CONNECT will
			# be used, followed by <method>. Otherwise, <method>
			# will be used for <request>.
			# [NOTE]: <method> is not case-sensitive.
			#
-O <sec>		# The amount of time (in seconds) to wait for output
			# to be sent via <proxy-server>. The timer is reset
			# each time data is sent via the socket. If specified,
			# the range is between: 0 - 120 seconds.
			# When omitted, defaults to: 10 seconds.
			#
-P <request-port>	# Forces <request-port> to override the default port
			# used where a scheme does not exist in <request>.
			# A request will default to:
			#	- 80 (HTTP:// scheme)
			#	- 443 (HTTPS:// scheme)
			#
-R <request-file>	# Proceses <request-file> containing domain(s)/URL(s)
			# to send via -s and -p arguments. If this option is
			# specified, it overrides -r argument as long as the
			# option appears before -r in the argument list. The
			# requests can be entered one per line and/or many
			# per line. In the latter case, they should follow
			# legal delimiters (see: -r argument).
			#
-S <scheme>		# The scheme to use for <request>. It must be one of
			# HTTP:// or HTTPS://. If a scheme is present within
			# <request>, then <scheme> is ignored. Otherwise,
			# <scheme> is applied to domain/URL in <request>.
			# [NOTE]: <scheme> is not case-sensitive.
			#
-V <verbosity-level>	# The verbosity level for output. It must be one of
			#	- 1 (normal output and request payload only)
			#	- 2 (normal output and response payload only)
			#	- 3 (combines #1 and #2)
			# [NOTE]: Requests/Responses are sent to stderr
			# [NOTE]: When -L is used, they will be directed to
			# <request-log>; redirect stderr if needed.
			#

[Arguments]

-h			# Output this menu and exit.
			#
-p <proxy-port>		# The proxy server's listening TCP port.
			#
-r [<request ...>]	# Domain(s)/URL(s) to send via -s and -p arguments. If
			# you decide to send multiple requests, they can be
			# enclosed in quotes and delimted by:
			#	- commas
			#	- pipes
			#	- semi-colons
			#	- spaces
			# When <request> is omitted, requests are retrieved
			# via standard input. This can be achieved with two
			# main approaches:
			#	1. Manually entering via this program's stdin
			#	2. Piping stdout into this program's stdin
			# For approach #1, requests can be manually entered:
			#	A. Several on one line (following delimiters)
			#	B. One per line
			# The above can be combined as well. Sending an EOF
			# (end-of-file) will complete the input and allow
			# for further processing of requests.
			# For approach #2, requests can be sent via stdout of
			# one process to the stdin of this tool. For example:
			#	A. echo "request1...requestN" | proxycheck ...
			#	B. cat request.txt | proxycheck ...
			# [Note]: For approach 2B, requests can be processed,
			# following the same rules as approach #1A and/or #1B.
			# [Note]: If <request> exists and either approach #1 or
			# #2 is attempted, <request> takes precedence and other
			# request processing methods are ignored.
			# [Note]: If -R option is used and it appears before
			# this argument, then -R takes priority.
			#
-s <proxy-server>	# The proxy server to send requests through.
			#
-v			# Output current program version and exit.
			#
```
## Examples

Connecting to a proxy (localhost), listing on port 8080/TCP and checking "www.foobar.com": 
```
proxycheck -s localhost -p 8080 -r www.foobar.com

...
Sun Jan 20 19:40:56 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1: status:400
...
```
Notice the result was a status of HTTP 400. This can happen when a Host header is required by the server, such as when using HTTP/1.1 for example. By adding the Host header, we can see a different result:
```
proxycheck -s localhost -p 8080 -r www.foobar.com -H h

...
Sun Jan 20 19:45:25 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1: status:200
...
```
Another example showing multiple requests. This example is using the HEAD method, with various delimiters. We also add the Accept, Connection, User-Agent, and Host headers:
```
proxycheck -H acuh -s localhost -p 8080 -r "www.foobar.com; http://www.example.com/some/page.html, www.google.com|www.bing.com;github.com"

...
Sun Jan 20 19:54:12 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1: status:400
Sun Jan 20 19:54:12 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=www.example.com: port=80: path=/some/page.html: ver=HTTP/1.1: status:501
Sun Jan 20 19:54:12 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=www.google.com: port=80: path=/: ver=HTTP/1.1: status:405
Sun Jan 20 19:54:12 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=www.bing.com: port=80: path=/: ver=HTTP/1.1: status:200
Sun Jan 20 19:54:12 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=github.com: port=80: path=/: ver=HTTP/1.1: status:301
...
```
We received some status warnings such as HTTP 405 and even a HTTP 501 - This example shows the previous execution, but we turn on verbosity level 3 so we can see the requests and the responses:
```
...
Sun Jan 20 19:57:05 2019: proxycheck: STANDARD: dst=localhost: dport=8080: request:
head http://www.foobar.com/ HTTP/1.1
Host: www.foobar.com
User-Agent: proxycheck
Accept: */*
Connection: close



Sun Jan 20 19:57:05 2019: proxycheck: STANDARD: dst=localhost: dport=8080: response:
HTTP/1.1 400 Bad Request
Server: nginx/1.14.1
Date: Mon, 21 Jan 2019 02:57:05 GMT
Content-Type: text/html
Content-Length: 173
Connection: close
...
Sun Jan 20 19:57:05 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1: status:400
Sun Jan 20 19:57:05 2019: proxycheck: STANDARD: dst=localhost: dport=8080: request:
head http://www.example.com/some/page.html HTTP/1.1
Host: www.example.com
User-Agent: proxycheck
Accept: */*
Connection: close



Sun Jan 20 19:57:05 2019: proxycheck: STANDARD: dst=localhost: dport=8080: response:
HTTP/1.0 501 Not Implemented
Content-Type: text/html
Content-Length: 357
Connection: close
Date: Mon, 21 Jan 2019 02:57:05 GMT
Server: ECSF (oxr/83C5)
...
Sun Jan 20 19:57:05 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=head: domain=www.example.com: port=80: path=/some/page.html: ver=HTTP/1.1: status:501
Sun Jan 20 19:57:05 2019: proxycheck: STANDARD: dst=localhost: dport=8080: request:
head http://www.google.com/ HTTP/1.1
Host: www.google.com
User-Agent: proxycheck
Accept: */*
Connection: close



Sun Jan 20 19:57:05 2019: proxycheck: STANDARD: dst=localhost: dport=8080: response:
HTTP/1.1 405 Method Not Allowed
Content-Type: text/html; charset=UTF-8
Referrer-Policy: no-referrer
Content-Length: 1589
Date: Mon, 21 Jan 2019 02:57:05 GMT
Connection: close
...
```
We can see that either the specififed method (HEAD) is not implemented or it is not allowed. We issue the check again but leverage the HTTP GET (which is the current default) instead of the HEAD method. Note: The verbosity level is omitted since we no longer need the level of detail:
```
proxycheck -H acuh -s localhost -p 8080 -r "www.foobar.com; http://www.example.com/some/page.html, www.google.com|www.bing.com;github.com"

...
Sun Jan 20 20:00:55 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1: status:200
Sun Jan 20 20:00:55 2019: proxycheck: WARN: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=www.example.com: port=80: path=/some/page.html: ver=HTTP/1.1: status:404
Sun Jan 20 20:00:56 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=www.google.com: port=80: path=/: ver=HTTP/1.1: status:200
Sun Jan 20 20:00:56 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=www.bing.com: port=80: path=/: ver=HTTP/1.1: status:200
Sun Jan 20 20:00:56 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=github.com: port=80: path=/: ver=HTTP/1.1: status:301
...
```
In this example, there is a slower response from the server. The default timeout per response is 10 seconds, but this can still be undesireable if we have a lot of URLs/domains to process:
```
proxycheck -s localhost -p 8080 -H huac -R test-urls.txt

...
Sun Feb  3 13:25:16 2019: proxycheck: validation complete - starting
Sun Feb  3 13:25:16 2019: proxycheck: client [neo] > server [localhost/8080]
Sun Feb  3 13:25:16 2019: proxycheck: socket proto: TCP
Sun Feb  3 13:25:16 2019: proxycheck: include header=Accept
Sun Feb  3 13:25:16 2019: proxycheck: include header=Connection
Sun Feb  3 13:25:16 2019: proxycheck: include header=Host
Sun Feb  3 13:25:16 2019: proxycheck: include header=User-Agent
Sun Feb  3 13:25:16 2019: proxycheck: building request(s)
Sun Feb  3 13:25:16 2019: proxycheck: requests built: 27
Sun Feb  3 13:25:16 2019: proxycheck: request: processing
Sun Feb  3 13:25:26 2019: proxycheck: WARN: STANDARD: read=Resource temporarily unavailable: dst=localhost: dport=8080: action=readreq: verb=GET: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1
Sun Feb  3 13:25:36 2019: proxycheck: WARN: STANDARD: read=Resource temporarily unavailable: dst=localhost: dport=8080: action=readreq: verb=GET: domain=www.bar.com: port=80: path=/sample.html: ver=HTTP/1.1
Sun Feb  3 13:25:46 2019: proxycheck: WARN: STANDARD: read=Resource temporarily unavailable: dst=localhost: dport=8080: action=readreq: verb=GET: domain=baz.com: port=80: path=/: ver=HTTP/1.1
...
```
We can choose to speed processing up a bit and specify a shorter duration with "-I" - Here, we choose to wait for 2 seconds before moving on:
```
proxycheck -s localhost -p 8080 -H huac -R test-urls.txt -I 2

...
Sun Feb  3 13:30:02 2019: proxycheck: validation complete - starting
Sun Feb  3 13:30:02 2019: proxycheck: client [neo] > server [localhost/8080]
Sun Feb  3 13:30:02 2019: proxycheck: socket proto: TCP
Sun Feb  3 13:30:02 2019: proxycheck: include header=Accept
Sun Feb  3 13:30:02 2019: proxycheck: include header=Connection
Sun Feb  3 13:30:02 2019: proxycheck: include header=Host
Sun Feb  3 13:30:02 2019: proxycheck: include header=User-Agent
Sun Feb  3 13:30:02 2019: proxycheck: building request(s)
Sun Feb  3 13:30:02 2019: proxycheck: requests built: 27
Sun Feb  3 13:30:02 2019: proxycheck: request: processing
Sun Feb  3 13:30:04 2019: proxycheck: WARN: STANDARD: read=Resource temporarily unavailable: dst=localhost: dport=8080: action=readreq: verb=GET: domain=www.foobar.com: port=80: path=/: ver=HTTP/1.1
Sun Feb  3 13:30:06 2019: proxycheck: WARN: STANDARD: read=Resource temporarily unavailable: dst=localhost: dport=8080: action=readreq: verb=GET: domain=www.bar.com: port=80: path=/sample.html: ver=HTTP/1.1
Sun Feb  3 13:30:08 2019: proxycheck: WARN: STANDARD: read=Resource temporarily unavailable: dst=localhost: dport=8080: action=readreq: verb=GET: domain=baz.com: port=80: path=/: ver=HTTP/1.1
...
```
In this example, we are testing two sites entered manually. Whem we are done, we send an EOF - Take note of the HTTP status responses returned:
```
proxycheck -s localhost -p 8080 -H huac -I 2 -r
bar.com
baz.com
<EOF Given Here>
Sun Feb  3 13:33:57 2019: proxycheck: validation complete - starting
Sun Feb  3 13:33:57 2019: proxycheck: client [neo] > server [localhost/8080]
Sun Feb  3 13:33:57 2019: proxycheck: socket proto: TCP
Sun Feb  3 13:33:57 2019: proxycheck: include header=Accept
Sun Feb  3 13:33:57 2019: proxycheck: include header=Connection
Sun Feb  3 13:33:57 2019: proxycheck: include header=Host
Sun Feb  3 13:33:57 2019: proxycheck: include header=User-Agent
Sun Feb  3 13:33:57 2019: proxycheck: building request(s)
Sun Feb  3 13:33:57 2019: proxycheck: requests built: 2
Sun Feb  3 13:33:57 2019: proxycheck: request: processing
Sun Feb  3 13:33:58 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=bar.com: port=80: path=/: ver=HTTP/1.1: status:301
Sun Feb  3 13:33:58 2019: proxycheck: STANDARD: dst=localhost: dport=8080: action=procresp: verb=GET: domain=baz.com: port=80: path=/: ver=HTTP/1.1: status:302
Sun Feb  3 13:33:58 2019: proxycheck: processing complete
Sun Feb  3 13:33:58 2019: proxycheck: cleaning up
Sun Feb  3 13:33:58 2019: proxycheck: finished
```
If we decided that we want to follow these redirects, we can use "-F" and specify the status codes we want to follow. Note: the result of "unable to follow" in each of the responses:
```
proxycheck -s localhost -p 8080 -H huac -I 2 -r -F 301,302
bar.com
baz.com
<EOF Given Here>
Sun Feb  3 13:38:04 2019: proxycheck: validation complete - starting
Sun Feb  3 13:38:04 2019: proxycheck: client [neo] > server [localhost/8080]
Sun Feb  3 13:38:04 2019: proxycheck: socket proto: TCP
Sun Feb  3 13:38:04 2019: proxycheck: include header=Accept

CHANGELOG.txt
Version 0.0.2

- Fixed issue when a domain is not present in a Location header
- Fixed core dump when attempting to process a linked list node that is NULL
- Preferred headers passed via argument list reflect properly when debugging
- Reimplemented buildreq() and fixed numerous logic issues
- HTTP Version can be HTTP 1.0 or HTTP 1.1 (defaults to 1.1)
- Overriding ports works as intended now
- Enhanced output to stdout/stderr and log
- Fixed abort trap when following an HTTP 3xx and Location header contained an 
appended port and new request but now built properly
- Redirects limited to avoid circular references (ex: www.bilibili.com); can 
override with "NO_LIMIT_MAX_REDIRECTS" directive
- Timestamps consistently applied
- Debugging directive outputs enhanced
- Fixed issue where -S was not being treated as case-insenstive
- Misc clean up / enhancements / fixes

Version 0.0.1

- Initial release

proxycheck.h
/*
    Purpose:
      Header file for "proxycheck".
    Developer:
      Ray Daley (https://github.com/darkestb4light)
   Note:
      - License: GNU General Public License Version 3
      | This file is part of proxycheck.
      |
      | proxycheck is free software: you can redistribute it and/or modify
      | it under the terms of the GNU General Public License as published by
      | the Free Software Foundation, either version 3 of the License, or
      | (at your option) any later version.
      |
      | proxycheck is distributed in the hope that it will be useful,
      | but WITHOUT ANY WARRANTY; without even the implied warranty of
      | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      | GNU General Public License for more details.
      |
      | You should have received a copy of the GNU General Public License
      - Used with:
      |__ proxycheck_main.c, proxycheck.c
*/

#ifndef PROXYCHECK_H
#if defined(__FreeBSD__) || defined(__gnu_hurd__)
#define _WITH_GETLINE
#elif defined(__GNU__) || defined(__linux__)
#define _GNU_SOURCE                             /* Don't modify */
#include <strings.h>                            /* Don't modify */
#elif defined(__APPLE__) || defined(__MACH__)
  /* intentionally empty */
#elif defined(__CYGWIN__) && !defined(_WIN32)
#define _GNU_SOURCE                             /* Don't modify */
#include <strings.h>
#include <sys/select.h>
#else
#error Target OS unsupported. Aborting.
#error If this is unexpected, submit an issue to the project:
#error https://github.com/darkestb4light/proxycheck
#endif           

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/types.h>

#ifdef AVOID_MALLOC_FREE                        /* Set if getting errors 
                                                   relating to malloc() or
                                                   free() and submit a bug 
                                                   to the project  
                                                */
#define AVOID_MALLOC_FREE           1
#else
#define AVOID_MALLOC_FREE           0
#endif

#ifdef SOCK_PROTO_IP                            /* Request socket */
#define SOCK_PROTO                  0
#else
#define SOCK_PROTO                  6
#endif

#ifndef NO_LIMIT_MAX_REDIRECTS                  /* Follow redirects "infinitely"
                                                   or set a maximum before 
                                                   giving up
                                                 */
#define MAX_REDIRECTS               3
#else
#define MAX_REDIRECTS               0
#endif


/* Debug definitions */

#ifdef DEBUG_BUILDREQ                           /* Show built requests */
#define DEBUG_BUILDREQ              1
#else
#define DEBUG_BUILDREQ              0
#endif

#ifdef DEBUG_PROCRESP                           /* Show processed responses */
#define DEBUG_PROCRESP              1
#else
#define DEBUG_PROCRESP              0
#endif

#ifdef DEBUG_READREQ                            /* Show responses received */
#define DEBUG_READREQ               1
#else
#define DEBUG_READREQ               0
#endif

#ifdef DEBUG_WRITEREQ                           /* Show requests sent */
#define DEBUG_WRITEREQ              1
#else
#define DEBUG_WRITEREQ              0
#endif

#ifdef DEBUG_FORCE_EXIT                         /* Force exit with each debug */
#define DEBUG_FORCE_EXIT            1
#else
#define DEBUG_FORCE_EXIT            0
#endif

#ifdef EXIT_CONN_WRITEREQ_ERR                   /* CONNECT request fail */
#define EXIT_CONN_WRITEREQ_ERR      1
#else
#define EXIT_CONN_WRITEREQ_ERR      0
#endif

#ifdef EXIT_CONN_READREQ_ERR                    /* CONNECT response fail */
#define EXIT_CONN_READREQ_ERR       1
#else
#define EXIT_CONN_READREQ_ERR       0
#endif

#ifdef EXIT_CONN_PROCRESP_ERR                   /* CONNECT parse fail */
#define EXIT_CONN_PROCRESP_ERR      1
#else
#define EXIT_CONN_PROCRESP_ERR      0
#endif

#ifdef EXIT_WRITEREQ_ERR                        /* STANDARD request fail */
#define EXIT_WRITEREQ_ERR           1
#else
#define EXIT_WRITEREQ_ERR           0
#endif     

#ifdef EXIT_READREQ_ERR                         /* STANDARD response fail */
#define EXIT_READREQ_ERR            1
#else
#define EXIT_READREQ_ERR            0
#endif 

#ifdef EXIT_PROCRESP_400_ERR                    /* STANDARD HTTP 400 fail */
#define EXIT_PROCRESP_400_ERR       1
#else
#define EXIT_PROCRESP_400_ERR       0
#endif 

#ifndef NO_EXIT_PROCRESP_407_ERR                /* STANDARD HTTP 407 fail */
#define EXIT_PROCRESP_407_ERR       1
#else
#define EXIT_PROCRESP_407_ERR       0
#endif
         
#ifndef NO_EXIT_PROCRESP_DEF_ERR                /* STANDARD unknown fail */
#define EXIT_PROCRESP_DEF_ERR       1
#else
#define EXIT_PROCRESP_DEF_ERR       0
#endif

#ifdef OUTPUT_REQ                               /* See request payload; 
                                                   REQ_BYTE_MAX might need 
                                                   to be modified to see 
                                                   desired results
                                                */
#define OUTPUT_REQ                  1
#else
#define OUTPUT_REQ                  0
#endif

#ifdef OUTPUT_RESP                              /* See request payload; 
                                                   RESP_SZ_MAX might need to 
                                                   be modified to see desired 
                                                   results
                                                */	
#define OUTPUT_RESP                 1
#else
#define OUTPUT_RESP                 0
#endif

#ifdef NO_CONN_FORCE_GET_DEF_VERB               /* Avoid CONNECT forcing GET
                                                   for default method;
                                                   Ignored when CONNECT is 
                                                   passed as an option
                                                */
#define NO_CONN_FORCE_GET_DEF_VERB  1
#else
#define NO_CONN_FORCE_GET_DEF_VERB  0
#endif

/* General definitions */

#define NAME                    "proxycheck"
#define VERSION                 "0.0.2"         /* Current version */
#define VER_BUF                 15              /* Version buffer */      
#define NAME_SZ                 11              /* Should match NAME len */
#define MAX_ARGS                17              /* Total args/opts */
#define REQ_ARG_OFFSET          13              /* Offset for required args */
#define DAEMON_DIR_DEF          "/"             /* Default path; If running
                                                   native on Windows, this 
                                                   path will be:
                                                   <drive>:\Users\<user>\ 
                                                */
#define DAEMON_SLEEP_MIN        1               /* Minimum seconds to sleep */
#define DAEMON_SLEEP_MAX        4294967295      /* Maximum seconds to sleep */
#define DAEMON_SLEEP_DEF        3600            /* Default seconds to sleep */
#define DAEMON_LOG_DEF          NAME ".log"     /* Default log file */
#define REQ_DEF_VERB            "GET"           /* Default method/verb */
#define FILE_BUF                2048            /* File buffer */
#define LOG_LEVEL_MIN           1               /* Minimum verbosity level */
#define LOG_LEVEL_MAX           3               /* Maximum verbosity level */
#define REQ_DEF_SCHEME          "http://"       /* Default scheme */
#define LOG_IONBF               1               /* Unbuffered (default) */
#define LOG_IOLBF               2               /* Line buffered */
#define LOG_IOFBF               3               /* Fully buffered */
#define LOG_STREAM_STDOUT       stdout          /* STDOUT log stream */
#define LOG_STREAM_STDERR       stderr          /* STDERR log stream */		
#define REQ_BYTE_MAX            100000000       /* Request buffer; This can 
                                                   be adjusted based on how 
                                                   large to allocate for the 
                                                   request buffer; Note: 
                                                   setting too low can hinder 
                                                   the amount of requests that 
                                                   can be processed, while 
                                                   setting too high may affect 
                                                   available memory or possibly 
                                                   performance; Use caution. 
                                                   Default is 10MB; 
                                                   1MB == 1000000 
                                                */
#define REQ_TOKENS              " ,;|"          /* Request delimiters */
#define	TIME_SZ                 26              /* Timestamp buffer; 
                                                   Don't modify 
                                                */
#define MAX_VERBS               4               /* HTTP methods; Don't modify */
#define MAX_SCHEMES             2               /* HTTP schemes; Don't modify */
#define MAX_REDIRECT_TYPES      4               /* HTTP redirects; 
                                                   Don't modify 
                                                */
#define HOST_SZ                 256             /* Host buffer; Don't modify */
#define PORT_MIN                0               /* Port minimum */
#define PORT_MAX                65535           /* Port maximum */
#define WRITE_TM_MIN            0               /* Min seconds - don't modify */
#define READ_TM_MIN             0               /* Min seconds - don't modify */ 
#define WRITE_TM_DEF            10              /* Request seconds (default) */
#define READ_TM_DEF             10              /* Response seconds (default) */
#define WRITE_TM_MAX            120             /* Max seconds - don't modify */
#define READ_TM_MAX             120             /* Max seconds - don't modify */

/* Request definitions */

#define REQ_VERB_SZ             8               /* Request method buffer */
#define REQ_SCHEME_SZ           9               /* Request scheme buffer */
#define REQ_PORT_SZ             6               /* Request port buffer */
#define REQ_HTTP_PORT_DEF       80              /* Request def. HTTP port */
#define REQ_HTTPS_PORT_DEF      443             /* Request def. HTTPS port */
#define REQ_PATH_DEF            "/"             /* Request URI default */
#define REQ_PATH_SZ             (2080+HOST_SZ)  /* URL buffer per request */
#ifdef REQ_HTTP_VER1_0
#define REQ_HTTP_VER_DEF        "HTTP/1.0"      /* Request version 1.0 */
#else
#define REQ_HTTP_VER_DEF        "HTTP/1.1"      /* Request version 1.1 */
#endif
#define REQ_HTTP_VER_SZ         9               /* HTTP version buffer */
#define REQ_ACCEPT_HDR_DEF      "*/*"           /* Accept header default */

                                                /* Request header buffers */
#define REQ_HOST_HDR_SZ         (6+HOST_SZ+4)   /* |__ Host */
#define REQ_ACCEPT_HDR_SZ       16              /* |__ Accept */
#define REQ_USER_HDR_SZ         (12+NAME_SZ+4)  /* |__ User-Agent */
#define REQ_CONN_HDR_SZ         (18+HOST_SZ+4)  /* |__ Host */

                                                /* Request CONNECT buffer */
#define REQ_CONN_SZ             (REQ_VERB_SZ+REQ_SCHEME_SZ+HOST_SZ+\
                                REQ_PATH_SZ+REQ_HTTP_VER_SZ)
                                
                                                /* Request STANDARD buffer */
#define REQ_SZ                  (REQ_CONN_SZ+REQ_HOST_HDR_SZ+\
                                REQ_CONN_HDR_SZ)

/* Response definitions */

#define RESP_MSG_403            "Forbidden/Proxy denial"
#define RESP_MSG_407            "Proxy Authentication Required"
#define RESP_LOC_BUF            HOST_SZ         /* Response Location buffer */
#define RESP_PROC_SZ            32              /* Response field buffer */	
#define RESP_SZ_MIN             10              /* Minimum response buffer */	
#define RESP_SZ_DEF             512             /* Default response buffer */
#define RESP_SZ_MAX             65535           /* Maximum response buffer */

struct Sockhost{                                /* Hold client/server info */
    char c_name[HOST_SZ];
    char s_name[HOST_SZ];
    char s_ip[HOST_SZ];	
    int s_port;
};

struct requestopt {                             /* Request options */
    char path[REQ_PATH_SZ];
    char scheme[REQ_SCHEME_SZ];
    char tmpurlport[REQ_PORT_SZ];
    char ver[REQ_HTTP_VER_SZ];
    char verb[REQ_VERB_SZ];
    char *tmpurl;
    const char *hdr;
    int port;
    unsigned int do_accept_hdr;
    unsigned int do_conn_hdr;
    unsigned int do_host_hdr;
    unsigned int do_user_hdr;
    unsigned int opt_scheme;
};

struct request {                                /* Request object(s) */
    char accept_hdr[REQ_ACCEPT_HDR_SZ];
    char conn_hdr[REQ_CONN_HDR_SZ];
    char connreq[REQ_CONN_SZ];
    char connverb[REQ_VERB_SZ];
    char domain[HOST_SZ];
    char host_hdr[REQ_HOST_HDR_SZ];
    char httpreq[REQ_SZ];
    char path[REQ_PATH_SZ];
    char port[REQ_PORT_SZ];
    char scheme[REQ_SCHEME_SZ];
    char user_hdr[REQ_USER_HDR_SZ];
    char ver[REQ_HTTP_VER_SZ];
    char verb[REQ_VERB_SZ];
    unsigned int valid_request;
    struct request *next;
};

/* Prototypes */
  
void buildreq(struct request *, struct requestopt *);
int daemonize(pid_t *, char *);
int getaddr(const char *, char *, const int);
int getlocation(char *, unsigned int, char *, unsigned int, char *);
int getreqcount(struct request *);
void getreqline(char *, size_t, FILE *, char *, long);
int gettime(char *);
int initsock(int, int, int, struct sockaddr_in, socklen_t);
void logevents(char *, int, FILE *);
long optdaemon(const char **, unsigned int *);
int *optfollow(const char **, unsigned int *, int *);
void optheader(const char **, unsigned int *, struct requestopt *);
int optinput(const char **, unsigned int *);
void optlog(const char **, unsigned int *, char *);
int optoutput(const char **, unsigned int *);
int optport(const char **, unsigned int *);
int optproxyport(const char **, unsigned int *);
int optproxyserver(const char **, unsigned int *, char *);
unsigned int optrespbuffer(const char **, unsigned int *);
void optreq(const char **, unsigned int *, char *, long);
void optreqfile(const char **, unsigned int *, char *, long);
void optscheme(const char **, unsigned int *, char *);
void optverb(const char **, unsigned int *, char *);
int optverbosity(const char **, unsigned int *);
void optversion(char *);
int procresp(const char *, int);
ssize_t readreq(int, char *, size_t, unsigned int, unsigned int);
void stripline(char *);
void usage(void);
size_t writereq(int, char *, size_t, unsigned int);

#endif
